src/record_ts.h

```diff
@@ -0,0 +1,873 @@
+#endif
+
+        this->m_internal = internal;
+
+        this->m_recording.backend = internal->backend;
+
+        this->scope = internal->scope;
+    };
+
+    void barrier() override {
+        if (!paused()) {
+            uint32_t start = this->m_recording.dependencies.size();
+
+            Operation op;
+            op.type             = OpType::Barrier;
+            op.dependency_range = std::pair(start, start);
+            this->m_recording.operations.push_back(op);
+        }
+
+        scoped_pause();
+        return this->m_internal->barrier();
+    }
+
+    Task *launch(Kernel kernel, KernelKey *key, XXH128_hash_t hash,
+                 uint32_t size, std::vector<void *> *kernel_params,
+                 const std::vector<uint32_t> *kernel_param_ids) override {
+        if (!paused()) {
+            try {
+                record_launch(kernel, key, hash, size, kernel_params,
+                              kernel_param_ids);
+            } catch (...) {
+                record_exception();
+            }
+        }
+        scoped_pause();
+        return this->m_internal->launch(kernel, key, hash, size, kernel_params,
+                                        kernel_param_ids);
+    }
+
+    /// Fill a device memory region with constants of a given type
+    void memset_async(void *ptr, uint32_t size, uint32_t isize,
+                      const void *src) override {
+        if (!paused()) {
+            try {
+                record_memset_async(ptr, size, isize, src);
+            } catch (...) {
+                record_exception();
+            }
+        }
+        scoped_pause();
+        return this->m_internal->memset_async(ptr, size, isize, src);
+    }
+
+    /// Mask compression
+    uint32_t compress(const uint8_t *in, uint32_t size,
+                      uint32_t *out) override {
+        if (!paused()) {
+            try {
+                record_compress(in, size, out);
+            } catch (...) {
+                record_exception();
+            }
+        }
+        scoped_pause();
+        return this->m_internal->compress(in, size, out);
+    }
+
+    /// Compute a permutation to reorder an integer array into discrete groups
+    uint32_t mkperm(const uint32_t *values, uint32_t size,
+                    uint32_t bucket_count, uint32_t *perm,
+                    uint32_t *offsets) override {
+        if (!paused()) {
+            try {
+                record_mkperm(values, size, bucket_count, perm, offsets);
+            } catch (...) {
+                record_exception();
+            }
+        }
+        scoped_pause();
+        return this->m_internal->mkperm(values, size, bucket_count, perm,
+                                        offsets);
+    }
+
+    /// Perform a synchronous copy operation
+    void memcpy(void *dst, const void *src, size_t size) override {
+        jitc_log(LogLevel::Debug, "record(): memcpy(dst=%p, src=%p, size=%zu)",
+                 dst, src, size);
```

Comment:
Consistently remove or keep the logs

Reply:
Changed the logging a bit. Is it consistent enough now?